File: mobile/android/app/src/main/java/app/terra/explore/WalkingStreak.kt
package app.terra.explore
import java.time.LocalDate
import java.time.Instant
import java.time.ZoneId

object WalkingStreak {
    data class Restore(val day: String,val usedOn: String)
    data class Status(val count: Int,val remaining: Int,val canRestore: Boolean,val yesterday: String,val today: String,val walkedToday: Boolean)
    fun days(sessions: List<Session>): Set<String> = sessions.filter { it.mode.category==Mode.walk && it.endedAt!=null && it.duration()>=300 && it.distance()>0 }.map { Instant.ofEpochMilli(it.endedAt!!).atZone(ZoneId.systemDefault()).toLocalDate().toString() }.toSet()
    fun status(walked: Set<String>, restores: List<Restore>, today: LocalDate=LocalDate.now()): Status {
        val all=walked+restores.map { it.day }; val yesterday=today.minusDays(1).toString(); val used=restores.count { it.usedOn.take(7)==today.toString().take(7) }
        var cursor=if(today.toString() in all) today else today.minusDays(1); var count=0
        while(cursor.toString() in all) { count++; cursor=cursor.minusDays(1) }
        return Status(count,(3-used).coerceAtLeast(0),used<3 && yesterday !in all && today.minusDays(2).toString() in all,yesterday,today.toString(),today.toString() in walked)
    }
}

File: mobile/ios/Terra/WalkingStreak.swift
import Foundation

enum WalkingStreak {
    struct Status { var count: Int; var remaining: Int; var canRestore: Bool; var yesterday: String; var today: String; var walkedToday: Bool }
    static func key(_ date: Date, calendar: Calendar = .current) -> String {
        let c = calendar.dateComponents([.year,.month,.day], from: date)
        return String(format: "%04d-%02d-%02d", c.year!,c.month!,c.day!)
    }
    static func days(_ sessions: [TrackSession], calendar: Calendar = .current) -> Set<String> {
        Set(sessions.filter { $0.mode.category == .walk && $0.endedAt != nil && $0.duration() >= 300 && $0.distance > 0 }.map { key(Date(timeIntervalSince1970: $0.endedAt!/1000), calendar: calendar) })
    }
    static func status(walked: Set<String>, restores: [StreakRestore], now: Date = Date(), calendar: Calendar = .current) -> Status {
        let today = calendar.startOfDay(for: now), yesterday = calendar.date(byAdding: .day, value: -1, to: today)!
        let todayKey = key(today,calendar: calendar), yesterdayKey = key(yesterday,calendar: calendar)
        let all = walked.union(restores.map(\.day))
        let used = restores.filter { $0.usedOn.prefix(7) == todayKey.prefix(7) }.count
        var cursor = all.contains(todayKey) ? today : yesterday, count = 0
        while all.contains(key(cursor,calendar: calendar)) { count += 1; cursor = calendar.date(byAdding: .day, value: -1, to: cursor)! }
        let before = key(calendar.date(byAdding: .day, value: -1, to: yesterday)!,calendar: calendar)
        return Status(count: count, remaining: max(0,3-used), canRestore: used < 3 && !all.contains(yesterdayKey) && all.contains(before), yesterday: yesterdayKey, today: todayKey, walkedToday: walked.contains(todayKey))
    }
}
